sdk/core/azure-core/src/main/java/com/azure/core/implementation/RestProxy.java

```diff
@@ -521,12 +525,37 @@ private static Exception instantiateUnexpectedException(UnexpectedExceptionInfor
 
                 switch (paramCount) {
                     case 3:
-                        return ctor.newInstance(httpRequest, responseStatusCode, responseHeaders);
+                        return Mono.just(ctor.newInstance(httpRequest, responseStatusCode, responseHeaders));
                     case 4:
-                        return ctor.newInstance(httpRequest, responseStatusCode, responseHeaders, bodyAsObject);
+                        return Mono.just(ctor.newInstance(httpRequest, responseStatusCode, responseHeaders,
+                            bodyAsObject));
                     case 5:
-                        return ctor.newInstance(httpRequest, responseStatusCode, responseHeaders, bodyAsObject,
-                            response.getDecodedHeaders().block());
+                        return response.getDecodedHeaders()
+                            .defaultIfEmpty(NullObject.INSTANCE)
```

Comment:
Is there a case when `getDecodedHeaders` returns null? If not this statement can be removed.

Reply:
It could be. This is for deserializedHeaders which is not required to have for all clients.

Reply:
you could use `switchIfEmpty` operator and avoid using `NullObject` pattern

Reply:
```
response.getDecodedHeaders()
    .map(headers -> {
       // code to create an instance of response that takes headers
    })
    .switchIfEmpty( // code to create an instance of response that does not take header);
```

Reply:
With help with @anuchandy , get this piece of code use reactor way of handling null. Done.